mongoengine/queryset/transform.py

```diff
@@ -98,6 +99,15 @@ def query(_doc_cls=None, **kwargs):
                         value = value['_id']
 
             elif op in ('in', 'nin', 'all', 'near') and not isinstance(value, dict):
+                if isinstance(value, BaseDocument):
```

Comment:
If all of the above truly expect an iterable, then we should raise an error for *any* non-iterable, not just a `BaseDocument`.